src/apps/testapps/testH3CellAreaExhaustive.c

```diff
@@ -0,0 +1,173 @@
+    ba = H3_EXPORT(pointDistRads)(&b, &a);
+    t_assert(ab > 0, pos);
+    t_assert(ab == ba, comm);
+
+    ab = H3_EXPORT(pointDistKm)(&a, &b);
+    ba = H3_EXPORT(pointDistKm)(&b, &a);
+    t_assert(ab > 0, pos);
+    t_assert(ab == ba, comm);
+
+    ab = H3_EXPORT(pointDistM)(&a, &b);
+    ba = H3_EXPORT(pointDistM)(&b, &a);
+    t_assert(ab > 0, pos);
+    t_assert(ab == ba, comm);
+}
+
+/**
+ * Tests positivity of edge length calculation for the functions:
+ *
+ *      exactEdgeLengthRads
+ *      exactEdgeLengthKm
+ *      exactEdgeLengthM
+ *
+ * @param  edge  edge to compute the length of
+ */
+static void edge_length_assert(H3Index edge) {
+    char msg[] = "edge has positive length";
+
+    t_assert(H3_EXPORT(exactEdgeLengthRads)(edge) > 0, msg);
+    t_assert(H3_EXPORT(exactEdgeLengthKm)(edge) > 0, msg);
+    t_assert(H3_EXPORT(exactEdgeLengthM)(edge) > 0, msg);
+}
+
+/**
+ * Test that cell area calculations are positive for the functions:
+ *
+ *      cellAreaRads2
+ *      cellAreaKm2
+ *      cellAreaM2
+ *
+ * @param  cell  cell to compute the area of
+ */
+static void cell_area_assert(H3Index cell) {
+    char msg[] = "cell has positive length";
+
+    t_assert(H3_EXPORT(cellAreaRads2)(cell) > 0, msg);
+    t_assert(H3_EXPORT(cellAreaKm2)(cell) > 0, msg);
+    t_assert(H3_EXPORT(cellAreaM2)(cell) > 0, msg);
+}
+
+/**
+ * Apply a cell area calculation function to every cell on the earth at a given
+ * resolution, and check that it sums up the total earth area.
+ *
+ * @param  res        resolution of the cells
+ * @param  cell_area  callback to compute area of each cell
+ * @param  target     expected earth area in some units
+ * @param  tol        error tolerance allowed between expected and actual
+ */
+static void earth_area_test(int res, double (*cell_area)(H3Index),
+                            double target, double tol) {
+    double area = mapSumAllCells_double(res, cell_area);
+
+    t_assert(fabs(area - target) < tol,
+             "sum of all cells should give earth area");
+}
+
+SUITE(h3CellArea) {
+    TEST(haversine_distances) {
+        iterateAllUnidirectionalEdgesAtRes(0, haversine_assert);
+        iterateAllUnidirectionalEdgesAtRes(1, haversine_assert);
+        iterateAllUnidirectionalEdgesAtRes(2, haversine_assert);
+        iterateAllUnidirectionalEdgesAtRes(3, haversine_assert);
+    }
+
+    TEST(edge_length) {
+        iterateAllUnidirectionalEdgesAtRes(0, edge_length_assert);
+        iterateAllUnidirectionalEdgesAtRes(1, edge_length_assert);
+        iterateAllUnidirectionalEdgesAtRes(2, edge_length_assert);
+        iterateAllUnidirectionalEdgesAtRes(3, edge_length_assert);
+    }
+
+    TEST(cell_area_positive) {
+        iterateAllIndexesAtRes(0, cell_area_assert);
+        iterateAllIndexesAtRes(1, cell_area_assert);
+        iterateAllIndexesAtRes(2, cell_area_assert);
+        iterateAllIndexesAtRes(3, cell_area_assert);
+    }
+
+    TEST(cell_area_earth) {
```

Comment:
I don't see assertions for specific cells in this PR - could we add some tests for cell size compared to a known good value calculated using a different method?

Reply:
Good point! I'll add some tests like that.

Reply:
Added!